---
title: "Cancer methylome analysis. Protocol group 2."
author: 'Krank K., Kuchina M., Mendoza-Alcala A., Spatuzzi M.'
output: html_document

# Introduction

  DNA methylation is a well studied epigenetic mechanism which influences many processes in a cell including cell proliferation and differentiation but also development of common diseases such as cancer.   Studies show that methylomes of cancer cells have a different methylation pattern, especially in promoter regions, which indicates that gene expression is inhibited. (Stephen B Baylin 2005) Out team worked with the DNA methylation data collected from patients with acute myeloid leukemia (AML) and healthy people.The main goal of the project was to identify differentially methylated regions (DMRs) -  DNA sequences with different methylation status. This is the matter of scientific and clinical interest because reliable DMRs might help to discern diseased and healthy patients. In the following protocol we provide a comprehensive workflow of our project.     
 
# Data preprocessing

The data-set we had been working with contained a lot of irrelevant information. It was also partly incomplete - with many not available or missing values.  The coverage depth variated from unreliably low to suspiciously high (possibly resulted by PCR duplicates). A transfromation of heteroscedastic beta-values to homoscedastic M-values would be benefitial for direct applying of most statistical tests (Pan Du, Xiao Zhang, Chiang-Ching Huang, Nadereh Jafari, Warren A Kibbe, Lifang Hou and Simon M Lin 2010). Therefore, the data must have been cleaned and normalized to make further analysis possible.  

## Data clean-up

Goals: 

* remove regions with low or too high coverage 

* remove regions with high amount of missing values

Work process:

1. Data-set has been imported and separated into specific subgroups. 
  We defined `promoters_only` variable and splitted it into `beta` and `coverage` datasets for further operations.  

```
promoters <- data$promoters;
promoters_only <- promoters[,-c(1:10)]

beta <- promoters_only[,c(1:18)]
coverage<- promoters_only[,c(19:36)]
```
  
2.  Removing sex chromosomes.
  Several studies show that there are sex differences in methylation status. For example the second X-chromosome in females and the heterochromatic region of Y-chromosome are strongly methylated which does not correlate with any disease but is a normal effect. (Bernardino J1, Lombard M, Niveleau A, Dutrillaux B. 2000, Jingyu Liu , Marilee Morgan, Kent Hutchison, Vince D. Calhoun 2010) Therefore, correction for sex is necessary and X- and Y-chromosomes must be removed to avoid gender bias.
  
```
cover_nox <- coverage[-which(promoters$Chromosome == "chrX"), ]
cover_noxy <- cover_nox[-which(promoters$Chromosome == "chrY"), ]
rm(coverage, cover_nox)
```

3. Coverage thresholds setting.

 3.1 The upper threshold.
 
  The general distribution of coverage values was visualized in a `log_cover_mean` plot. On the x-axis we plotted mean values of coverage in a logarythmic scale while the y-axis shows the frequency of coverage depths in our data-set. With 5% and 95% quantiles we identified that 90% of the data is in the coverage range of 100x to 5500x. As expected there was no difference between coverage plot with and without sex chromosomes.
  
``` {r, echo=F, out.width="90%"}
data <- readRDS(file = "AML_gran_list.RDS")

promoters <- data$promoters;
promoters_only <- promoters[,-c(1:10)]
beta <- promoters_only[,c(1:18)]
coverage<- promoters_only[,c(19:36)]

cover_nox <- coverage[-which(promoters$Chromosome == "chrX"), ]
cover_noxy <- cover_nox[-which(promoters$Chromosome == "chrY"), ]
rm(coverage, cover_nox)

beta <- promoters_only[,c(1:18)]
coverage<- promoters_only[,c(19:36)]

cover_tot_mean <- rowMeans(cover_noxy) 
log_cover_mean <- log10(cover_tot_mean)


hist(log_cover_mean, breaks = 200, xlim = c(0,5.5), main="With XY chromosomes")
abline(v=quantile(log_cover_mean, probs = c(seq(0.1, 1.0, by= 0.1))), col="red")
abline(v=quantile(log_cover_mean, probs = c(0.05, 0.95), col="cyan"), lty=2)
hist(log_cover_mean, breaks = 200, xlim = c(0,5.5), main = "Whithout XY chromosomes")
abline(v=quantile(log_cover_mean, probs = c(seq(0.1, 1.0, by= 0.1))), col="red")
abline(v=quantile(log_cover_mean, probs = c(0.05, 0.95), col="cyan"), lty=2)
```

  After that we decided to create a plot which would show the amount of values we would lose depending on the threshold settings. We used the `unlist` function in order to freely inspect the whole dataframe. Here are the results for 75%, 95% and 100% quantiles:

``` {r, echo=F}
coverage_histo <- unlist(cover_noxy)
log_cover_histo <- unlist(log10(cover_noxy))
quantile(coverage_histo, probs=c(.75, .95, 1))
```
  We decided that the upper threshold should have been located between 95% and 100% quantiles. Vectors `NAs_up` and `thresholds_up` were created. The following loop goes through the data-set in 200-steps and looks for the values which coverage is higher than the threshold. 

```
NAs_up <- c()
thresholds_up <- c()
for(i in seq(0,length(coverage_histo),200)){
  threshold <- i
  thresholds_up <- append(thresholds_up, threshold)
  Unreliable <- coverage_histo[which(coverage_histo > threshold)]
  N_Unreliable <- length(Unreliable)
  NAs_up <- append(NAs_up, N_Unreliable)
}

```

```{r, echo=F}
NAs_up <- c()
thresholds_up <- c()
for(i in seq(0,length(coverage_histo),200)){
  threshold <- i
  thresholds_up <- append(thresholds_up, threshold)
  Unreliable <- coverage_histo[which(coverage_histo > threshold)]
  N_Unreliable <- length(Unreliable)
  NAs_up <- append(NAs_up, N_Unreliable)
}

par(mfrow=c(1,1))
plot(x=log10(thresholds_up), y=NAs_up, type= "l", main = "Unreliable Data by threshold", xlab= "Threshold position", ylab = "Unreliable Data")
abline(v=quantile(log_cover_mean, probs = c(seq(0.95, 1.0, by= 0.01))), col="green")
abline(v=quantile(log_cover_mean, probs = 0.995), col="orange")
abline(v=quantile(cover_tot_mean, probs = c(seq(0, 1.0, by= 0.1))), col="green")
quantile(coverage_histo, probs=.95)
```
The plot shows how much information would be transformed to `NA`s depending on the position of the threshold. On this basis we chose 95% quantile with the value 5888 as our upper threshold.

  3.2 The lower threshold
  
  For lower threshold settings we created vectors `NAs_low` and `thresholds_low`. Same operations as for the upper threshold were carried out.
  
```
NAs_low <- c()
thresholds_low <- c()
for(i in 1:35){
  threshold <- i
  thresholds_low <- append(thresholds_low, threshold)
  Unreliable <- coverage_histo[which(coverage_histo < threshold)]
  N_Unreliable <- length(Unreliable)
  NAs_low <- append(NAs_low, N_Unreliable)
}
```

```{r, echo=F}
NAs_low <- c()
thresholds_low <- c()
for(i in 1:35){
  threshold <- i
  thresholds_low <- append(thresholds_low, threshold)
  Unreliable <- coverage_histo[which(coverage_histo < threshold)]
  N_Unreliable <- length(Unreliable)
  NAs_low <- append(NAs_low, N_Unreliable)
}
par(mfrow=c(1,1))
plot( x=thresholds_low, y=NAs_low, type= "l", main = "Unreliable Data by threshold", 
      xlab= "Threshold position", ylab = "Unreliable Data")
```
  As shown on the plot the threshold position is proportional to the data we would lose. There is no "kink" as in the previous plot which could help up identify the upper threshold position. So we took the recommended value for the minimum coverage of 30x. (Ziller, M. J., Hansen, K. D., Meissner, A., & Aryee, M. J. (2015)

  3.3 Applying coverage thresholds to the data-set
  
  After establishing the upper and lower threshold positions we applied them to our data-set. New variable `promoters_qc` was created. All of the beta-values with coverage higher or lower than the thresholds were converted into `NA`.
  
```
lower_threshold <- 30
upper_threshold <- 5888

promoters_qc <- promoters_only
for(j in 19:36){
  for(i in 1:nrow(promoters_only)) {
    if(promoters_qc [i,j]  < lower_threshold | promoters_qc [i,j] > upper_threshold){
      promoters_qc [i,j-18] <- NA
    }
  }
}
```

  Genes with more than two `NA`s were then removed from both cancer and healthy cohort. We tolerated this amount of missing values because this way the data-sen would still be statistically significant.
(Yiran Dong and Chao-Ying Joanne Peng, 2013)

```
beta_can <- promoters_qc[,c(1:9)]
beta_con <- promoters_qc[,c(10:18)]

NAs_can <- rowSums(is.na(beta_can))
NAs_con <- rowSums(is.na(beta_con))

Can <- cbind(beta_can, NAs_can)
Con <- cbind(beta_con, NAs_con)

beta <- cbind(Can,Con)
beta_partly_cleaned <- beta[-which(beta[,10]> 2 |beta[,20] >2),]
```

  And finally the `NA`s were replaced with the mean value of the cohort (healthy or diseased) so it wouldn't affect further statistical tests.
  
```
for (i in 1:nrow(beta_partly_cleaned)) {
  n <- beta_partly_cleaned[i,10]
  
  if (n==1 | n==2) {
    for (j in 1:9){
      if (is.na(beta_partly_cleaned[i,j])){
        beta_partly_cleaned[i,j] <- rowMeans(beta_partly_cleaned[i,1:9], na.rm = T)
      }
    }
  }
}

beta_can_cleaned <- beta_partly_cleaned

for (i in 1:nrow(beta_can_cleaned)) {
  n <- beta_can_cleaned[i,20]
  
  if (n==1 | n==2) {
    for (j in 11:19){
      if (is.na(beta_can_cleaned[i,j])){
        beta_can_cleaned[i,j] <- rowMeans(beta_can_cleaned[i,11:19], na.rm = T)
      }
    }
  }
}
beta_ <- beta_can_cleaned[,-c(10,20)]
```


## Data normalization

Goal:

* transfer beta-values to more statistically valid M-values

Work process:
  
The M-value is calculated as the log2 ratio of the intensities of methylated probe versus unmethylated probe so here we normalized our data by transforming the beta-values with the following loop.

```
for (j in 1:18) {
  for (i in 1:nrow(beta_)) {
    beta_[[i,j]] <- log2(beta_[i,j] / (1-beta_[i,j]) )
    
  }
}

promotors_normalized <- beta_
```


# Statistical analysis



# Interpretation